### PR TITLE
Change the root password spoke GUI design

### PR DESCRIPTION
**Pull request by VladimirSlavik**
This resolves ambiguity of what options are available when root is locked.

~~This is a draft because there are at least these TODOs:~~
- [x] make bold labels "clickable" 
- [x] verify the logic
- [x] find bugs etc. to link
- [x] consider renaming spoke to "Root account security"
- [x] ~~add an "empty password" banner~~ automatically switches to no password + locked instead.

See also: https://github.com/rhinstaller/anaconda/pull/3262#issuecomment-839202230
Suggested-by: Mairin Duffy <mairin@redhat.com> @mairin

### CODE REVIEW

**Comment by VladimirSlavik** [comment]
@mairin, please comment on the preferred form of attribution, if you want it different.

**Comment by VladimirSlavik** [comment]
Here's how it looks.
(That shall not change substantially, anymore.)

Screenshot_vs_main_2021-07-14_16:24:05

Screenshot_vs_main_2021-07-14_16:24:12


https://user-images.githubusercontent.com/15903878/125641078-e552f276-c430-48a0-bdca-459347c0b3bf.mp4

**Comment by mairin** [comment]
Wow that looks wonderful, @VladimirSlavik !!! The padding / margins / spacing etc. are right on and the animation looks slick. 🎉

File: pyanaconda/ui/gui/spokes/root_password.py [hunk 1]
```diff
@@ -159,13 +160,16 @@ def refresh(self):
...
-        self._lock.set_active(self._users_module.IsRootAccountLocked)
+        if self._users_module.IsRootAccountLocked:
+            self._disable_radio.set_active(True)
+        else:
+            self._enable_radio.set_active(True)
         self._root_password_ssh_login_override.set_active(
             self._users_module.RootPasswordSSHLoginAllowed
         )
-        if not self._lock.get_active():
+        if self._enable_radio.get_active():
```
Comment (poncovka) [inline pyanaconda/ui/gui/spokes/root_password.py:172]: Maybe `self._enable_radio.get_active()` could be hidden in a new property.

File: pyanaconda/ui/gui/spokes/root_password.py [hunk 2]
```diff
@@ -179,7 +183,7 @@ def status(self):
         if self._users_module.IsRootAccountLocked:
             # reconfig mode currently allows re-enabling a locked root account if
             # user sets a new root password
-            if is_reconfiguration_mode() and not self._lock.get_active():
+            if is_reconfiguration_mode() and not self._enable_radio.get_active():
```
Comment (poncovka) [inline pyanaconda/ui/gui/spokes/root_password.py:186]: This needs to be tested with the initial setup.

**Review by poncovka — COMMENTED** [review]
It looks great!

**Comment by poncovka** [comment]
Please, test it with `rootpw --lock password`. I don't think we need to support this combination in UI, but the kickstart installation should still work. You can also run the `rootpw-lock` kickstart test.

**Comment by jstodola** [comment]
It looks like a great improvement in usability of the spoke!

I would just like to express my dislike for the animation. Although it looks OK when the installation is running locally, animations in the installer in general do not look good when running a remote installation via VNC, where you can see for example just one or two frames from the animation, depending on the network speed. In some case, animations make the installation process even painful, see https://bugzilla.redhat.com/show_bug.cgi?id=1946064. No animation is good animation.

**Comment by jstodola** [comment]
@AdamWill please be aware of this UI change, since it might affect your Fedora installation testing.

**Comment by poncovka** [comment]
Also, add a release note for this change, please (https://anaconda-installer.readthedocs.io/en/latest/contributing.html#release-notes).

**Comment by VladimirSlavik** [comment]
UI logic now works in absence of kickstart, onwards to that...

**Comment by mairin** [comment]
@VladimirSlavik Hey! Taking a look now - 

Re: attribution - looks fine to me :)

Re: alignments... if the change you made to make the alignment intentionally different (e.g. the added indent on the description for each radio button option) prevents the shifting of elements you were worried about, the alignment screenshot you posted looks fine to me! I did examine the GNOME HIG and it does not account for the case where radio buttons have a header message + a non-header description, so this doesn't explicitly violate any design patterns outlined there, and I think it looks fine & clear.

Re: the animation - is there any way you could give me a preview so I could see what it looks like? Probably fade in / fade out will look fine, I just want to see how it comes off!

**Comment by VladimirSlavik** [comment]
Thanks - and here you go!

https://user-images.githubusercontent.com/15903878/126683198-b5e887f1-f41e-450a-a551-6a883657eeb5.mp4

**Comment by mairin** [comment]
@VladimirSlavik looks great to me, I see no issue!

**Comment by jstodola** [comment]
The text boxes for entering the password are too wide - if they are as wide as possible, the screen would not look good on fullhd/4k monitors. Could the size of the password text boxes be reduced?
And one note about enabling/disabling remote access mentioned in the spoke - would it make sense to specify that it only affect remote access for the root account? Remote access for non-root users is not affected by this spoke.

**Comment by VladimirSlavik** [comment]
Thanks for the reminder, I forgot about that part at some point. I changed the width to be 40 characters, not as wide as possible.

I also changed the wording - it's a separate commit so it can be squashed or dropped according to feelings about that.

Screenshot_vs_main_2021-07-23_10:58:09

**Comment by VladimirSlavik** [comment]
I don't think we are tracking any Fedora bugs / issues to link. The RHEL 9 ones are technically not related until this is ported. So linking nothing.

**Comment by jstodola** [comment]
It looks great to me, thanks!

File: pyanaconda/ui/gui/spokes/root_password.py [hunk 3]
```diff
@@ -316,36 +316,32 @@ def _check_spoke_exit_conditions(self, unwaivable_check_failed):
...
-            # this is not a "real" click, just refresh() setting the lock check
-            # box state based on data from the DBus module
-            if not self._manually_locked:
-                # if the checkbox has not yet been manipulated by the user
-                # we can ignore this run and not adjust the fields
-                return True
-        self.password_entry.set_sensitive(not lock.get_active())
-        self.password_confirmation_entry.set_sensitive(not lock.get_active())
-        if not lock.get_active():
+    def on_disable_enable_clicked(self, control):
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/root_password.py:342]: This method name seems pretty confusing to me. Right now I'm confused if it should toggle radio buttons or disable some and enable some. Could you please change the name?

File: pyanaconda/ui/gui/spokes/root_password.py [hunk 4]
```diff
@@ -79,8 +77,10 @@ def initialize(self):
         self._password_confirmation_entry = self.builder.get_object("password_confirmation_entry")
         self._password_bar = self.builder.get_object("password_bar")
         self._password_label = self.builder.get_object("password_label")
-        self._lock = self.builder.get_object("lock")
+        self._enable_radio = self.builder.get_object("enable_radio")
+        self._disable_radio = self.builder.get_object("disable_radio")
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/root_password.py:81]: I would like to change these variable names. I'm missing the target here. What to enable root account, ssh access, password field?

**Review by jkonecny12 — CHANGES_REQUESTED** [review]
Hi, I really like the new solution, however, I would like to make the variable names a bit more obvious.

**Comment by poncovka** [comment]
> Thanks for the reminder, I forgot about that part at some point. I changed the width to be 40 characters, not as wide as possible.

I think you should limit the width of the whole action area to make it more centred.

File: pyanaconda/ui/gui/spokes/root_password.py (hunk 3 above)
Comment (VladimirSlavik) [reply pyanaconda/ui/gui/spokes/root_password.py:342]: Sorry, I don't agree. It's an event handler - the name conforms to the `on_controlname_eventtype` pattern most of our code uses. You do not call it, Gtk does. And the next line is a docstring, which also says what it is: _Click handler for enable and disable radio buttons._

**Comment by VladimirSlavik** [comment]
Fun fact of the day: GTK3+ does not have a way to set maximum width. Instead, you have to make sure nothing within the widget of interest will expand. In this case, it was simple, the only expanding thing were the labels. As luck has it, labels, unlike other controls, do have a "maximum width" property... set in characters. Then, the "minimal width" can be set by adding a width request to something. I feel like it's 2000 again and I am trying to vertically center a thing on the web...

Here is 60 characters:

Screenshot_vs_main_2021-07-29_22:02:06

File: pyanaconda/ui/gui/spokes/root_password.py (hunk 4 above)
Comment (VladimirSlavik) [reply pyanaconda/ui/gui/spokes/root_password.py]: I added "root" in the middle of the names, eg. disable_root_radio.

**Comment by VladimirSlavik** [comment]
Update: Limited width, rename radio buttons.

File: pyanaconda/ui/gui/spokes/root_password.py (hunk 3 above)
Comment (jkonecny12) [reply pyanaconda/ui/gui/spokes/root_password.py:342]: In that case is `disable_enable` a good control name?

Comment (VladimirSlavik) [reply pyanaconda/ui/gui/spokes/root_password.py:342]: The handler is used by two controls, one enables, one disables.

Maybe not mentioning the control names would make the name work? Eg. `on_root_enabled_changed`

Comment (jkonecny12) [reply pyanaconda/ui/gui/spokes/root_password.py:342]: That seems much better to understand to me.

**Comment by jkonecny12** [comment]
I don't agree with the current `enable_disable` but I don't think it's reason for blocking this PR.

**Review by jkonecny12 — COMMENTED** [review]
I don't agree with the current `enable_disable` but I don't think it's reason for blocking this PR.

File: pyanaconda/ui/gui/spokes/root_password.py (hunk 3 above)
Comment (VladimirSlavik) [reply pyanaconda/ui/gui/spokes/root_password.py:342]: Great, I will do that, then.

Comment (VladimirSlavik) [reply pyanaconda/ui/gui/spokes/root_password.py]: Done.

**Comment by VladimirSlavik** [comment]
Update: Changed the name of the handler per @jkonecny12 suggestion - thank you!

**Review by jkonecny12 — APPROVED** [review]
Looks great to me now. Thanks!

**Review by poncovka — APPROVED** [review]

**Comment by VladimirSlavik** [comment]
Update: Renamed to Root Account. The longer version with Security seems to still fit in English, but I wonder about all the translations.

**Review by poncovka — APPROVED** [review]
LGTM!

**Review by jkonecny12 — APPROVED** [review]
Still looks good to me!

**Comment by VladimirSlavik** [comment]
@AdamWill this will break OpenQA, and #3538 too. I'm about to merge these so the breakage is imminent.

(While at it, I'd like to also get into next Fedora also some rehash of #2242. That would address the easier GUI wants. But that's only hopes so far.)

**Pull request merged**